[ZEPPELIN-1357,1892,1370][Umbrella] Text overlap in the MultiBarChart

soralee opened the pull request:
### What is this PR for?
If using the `multiChartBar`, it often occurs overlap in text. So, this PR is for preventing to overlap in text. 

>~~The below is the case what I updated.~~
~~1. The xLabel generally show all of the text without rotation (less than 30).~~
~~2. If the xLabel size is over than 30 and less than 80, the xLabel text are rotated as 90 degree and displayed.~~
~~3. If the xLabel size is over than 80, the xLabel text are disappeared, but the tooltip is displayed.~~
>~~I have made improvements based on my thinking, so I would appreciate to give me your feedback.~~
~~And if merged this PR, I'll update the feature so that user can use `min` and `max` variables by user on web.~~

**[Update]**
To prevent overlap in xLabel text, in one way which is talked the below mentions, user who is using zeppelin could select to visualization type of xLabel such as `rotate 45 degree` or `hide`.

### What type of PR is it?
[Bug Fix | Improvement ]

### What is the Jira issue?
* [ZEPPELIN-1357; UI - label wrapping not done properly in charts](https://issues.apache.org/jira/browse/ZEPPELIN-1357)
* [ZEPPELIN-1892; Display label vertically or horizontally smartly](https://issues.apache.org/jira/browse/ZEPPELIN-1892)
* [ZEPPELIN-1370; Label overlaps - in default visualization example barchart](https://issues.apache.org/jira/browse/ZEPPELIN-1370)

### How should this be tested?
1. Run the paragraph for bank data in  the `Basic Features (Spark)` notebook.
2. Execute the following query.
``` 
%sql
select * from bank
```
3. Use the pivot in the `Setting` toggle in the `MultiBarChart` such as the screenshots.

### Screenshots (if appropriate)
**[Before]**
![z_zeppelin-1357_b](https://cloud.githubusercontent.com/assets/8110458/23898303/f80d7bbc-08f3-11e7-9b24-3248c492b8af.png)

![z_1357_b3_](https://cloud.githubusercontent.com/assets/8110458/23898435/75b3ae60-08f4-11e7-9d7c-31746f0d4edc.png)

**[After]**
![peek 2017-04-17 14-58](https://cloud.githubusercontent.com/assets/8110458/25080820/a3bfc1c0-2381-11e7-9c2b-16c0aa71234e.gif)






### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

Leemoonsoo commented on zeppelin-web/src/app/visualization/builtins/visualization-nvd3chart.js:
```diff
@@ -37,14 +37,39 @@ export default class Nvd3ChartVisualization extends Visualization {
 
+    // prevent to overlap in text if using multiBarChart
+    // if size of xLabel are over maxRotate, the MutiBarChart don't show text
+    if (type === 'multiBarChart' && d3g[0]) {
```
Could you try place this routine inside of https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/visualization/builtins/visualization-barchart.js#L58 ?

soralee replied:
OK, I'll try do it! Thanks :slightly_smiling_face:

soralee replied:
I updated it!

soralee commented:
@Leemoonsoo I really thanks for your feedback and your think! :blush: 
You mean, do you think it would be better to show `min` and `max` values ​​without `rotate` and `hide`?
Sounds good! Let's look at the opinions of others and change `min` and `max` instead of `rotate` and `hide`, if they want.

And here is my comment about your sayings and why I used the `rotate` and `hide`. 
>IMO, if bar chart automatically calculate and show label without overlap without user parameter that would be optimal.

Right! At first, I had also wanted that bar chart can be automatically calculate and show label without overlap and user parameter. However, because the nvd3 don't have These functions (or I might miss it), in the easy way, many users are using rotation for avoid overlap.
But, if there is a large amount of data, this rotation way also has a limit to showing label without overlap, so I handled by hiding the label.

Leemoonsoo commented:
> You mean, do you think it would be better to show min and max values ​​without rotate and hide?
Sounds good! Let's look at the opinions of others and change min and max instead of rotate and hide, if they want.

Sorry for confuse you. what i meant was more like select either `rotate` and `hide` is more easy to understand than set number for `min`, `max`.

soralee commented:
Thanks for quick response! 

> Sorry for confuse you. what i meant was more like select either rotate and hide is more easy to understand than set number for min, max.

Okay! I got it! 
Let's see others opinion :blush: 
I think `hide` is better! because this way can completely prevent overlap.

soralee commented:
I think the following JIRA issues are similar with this PR
* [ZEPPELIN-1892; Display label vertically or horizontally smartly](https://issues.apache.org/jira/browse/ZEPPELIN/ZEPPELIN-1892)
* [ZEPPELIN-1370; Label overlaps - in default visualization example barchart](https://issues.apache.org/jira/browse/ZEPPELIN-1370)

Leemoonsoo commented:
@soralee can you rebase or merge master and see if ci becomes green?

soralee commented:
Thanks @Leemoonsoo! I did it and CI is green now

1ambda commented on zeppelin-web/src/app/visualization/builtins/visualization-barchart.js:
```diff
@@ -57,6 +59,29 @@ export default class BarchartVisualization extends Nvd3ChartVisualization {
+    // if size of xLabel are over maxRotate, the MutiBarChart don't show text
+    var minRotate = 30;
+    var maxRotate = 80;
+    var xLabelsLength = Object.keys(self.xLabels).length;
```
MINOR: `self.xLabels` looks like an array, do we need `Object.keys` here?

soralee replied:
You're right, I missed when moving the code!
Let me check and update again, Thanks @1ambda

soralee replied:
Hi, @1ambda 
I check data type of xLabels was object because `d3Data.xLabels` is object. 
so, I removed Line 26 that is not necessary code.

1ambda replied:
great.

Leemoonsoo commented:
I tested a bit more and found some cases.

1. Label rotated when unnecessary
![image](https://cloud.githubusercontent.com/assets/1540981/24431884/7500d71a-13d3-11e7-9a0e-1b4514f86a09.png)

2. Label still overlapping when paragraph width is small
![image](https://cloud.githubusercontent.com/assets/1540981/24431897/8c620ffa-13d3-11e7-8f40-3f6e53ffbb63.png)

To address these problem, I think rotation/hide should be calculated based on actual chart width and combined width of all labels.
Or just providing completely manual (provide option to select normal/rotation/hide) could be alternative.

@soralee what do you think? Do you have any other idea?

soralee commented:
@Leemoonsoo Thanks for review and suggestions!
I think manual options depending on users seems great so far, because I didn't find automatic calculation about width of `xLabel`. Let me dig this part more and leave a comment after updating.

soralee commented:
@Leemoonsoo I updated description and code as that we talked about above comment. I didn't found how to be calculated based on actual width changed so, I updated manual button it can be used by user wants. 
Could you review again, please?

Leemoonsoo commented:
Tried last commit and it looks great.

However, in the beginning "Default" radio button is not selected automatically.
![image](https://cloud.githubusercontent.com/assets/1540981/24997221/cc29889e-2070-11e7-9e67-3c27b46095bb.png)

Also it'll be helpful to have a label for these radio buttons. Such as 'xaxis' or something.

soralee commented:
@Leemoonsoo Thanks for review!
Right! Label would be better :+1: 
Let me update again.

soralee commented:
@Leemoonsoo I added label and default variable furthermore, I changed from radio to button like this.
![peek 2017-04-17 14-58](https://cloud.githubusercontent.com/assets/8110458/25080413/dea68ce0-237e-11e7-929d-d166a769bb9c.gif)

Leemoonsoo commented:
Great improvement @soralee. Tested and LGTM.
Merge to master if no further discussions.

Leemoonsoo commented:
Ah, just one more thing. @soralee could you make CI green?

soralee commented:
@Leemoonsoo Yeah, this CI problem is related in #2255 
if merging #2255, I'll rebase this PR!

soralee commented:
@Leemoonsoo Ready to merge!

Leemoonsoo commented:
Thanks @soralee. merge to master.
